### PR TITLE
Rust

### PR DESCRIPTION
**Pull request by jravetch**
Thanks to the work of @Tinche Rust can now be added as a framework.  I've updated the download url and few tweaks to get Rust installed properly.  Addresses #104

### CODE REVIEW

File: tests/large/test_rust.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+
```
Comment (didrocks) [inline tests/large/test_rust.py]: Extra empty line here

File: tests/large/test_rust.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+# -*- coding: utf-8 -*-
+# Copyright (C) 2014 Canonical
```
Comment (didrocks) [inline tests/large/test_rust.py]: Maybe add your name + update year like 2014-2015?

File: tests/large/test_rust.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#
+# Authors:
+#  Didier Roche
+#  Tin Tvrtković
```
Comment (didrocks) [inline tests/large/test_rust.py]: Maybe add your name here (same for other files)

File: tests/large/test_rust.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            compile_command = ["bash", "-l", "rustc /tmp/hello.rs"]
+
+        self.child = spawn_process(self.command('{} rustc'.format(UMAKE)))
+        self.expect_and_no_warn("Choose installation path: {}".format(self.installed_path))
+        self.child.sendline("")
+        self.expect_and_no_warn("Installation done", timeout=self.TIMEOUT_INSTALL_PROGRESS)
+        self.wait_and_close()
+
+        self.assert_exec_exists()
+        self.assertTrue(self.is_in_path(self.exec_path))
```
Comment (didrocks) [inline tests/large/test_rust.py]: As you are also adding another directory in cargo/bin, I would also assertTrue that one binary there exists + is in path to ensure both paths were added.

File: tests/large/test_rust.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        # compile a small project
+        output = subprocess.check_output(self.command_as_list(compile_command)).decode()
+
+        if self.in_container:
+            self.assertEqual(output, "hello, world\r\n")
+        else:
+            self.assertEqual(output, "hello, world")
```
Comment (didrocks) [inline tests/large/test_rust.py]: Are you sure there is no trailing \n here but \r\n in the container?

File: tests/medium/test_rust.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+"""Tests for rust"""
+
+from . import ContainerTests
+import os
+from ..large import test_rust
+
+
+class RustInContainer(ContainerTests, test_rust.GoTests):
```
Comment (didrocks) [inline tests/medium/test_rust.py]: I think it should inherit from test_ruest.RustTests and not GoTests? (did you run medium tests? I guess not :p)

File: umake/frameworks/rust.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    def get_metadata_and_check_license(self, result):
+        """Override this so we can use BS and fetch the checksum separately."""
+        logger.debug("Fetched download page, parsing.")
+
+        page = result[self.download_page]
+
+        error_msg = page.error
+        if error_msg:
+            logger.error("An error occurred while downloading {}: {}".format(self.download_page_url, error_msg))
+            UI.return_main_screen()
```
Comment (didrocks) [inline umake/frameworks/rust.py]: Should pass 1 in parameter for exit code here.

File: umake/frameworks/rust.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        link = (soup.find('div', class_="install")
+                .find('td', class_="inst-type", text="Linux (.tar.gz)")
+                .parent
+                .find(text=self.arch_trans[get_current_arch()])
+                .parent
+                .parent)
+
+        if link is None:
+            logger.error("Can't parse the download URL from the download page.")
+            UI.return_main_screen()
```
Comment (didrocks) [inline umake/frameworks/rust.py]: Same here, pass 1.

File: umake/frameworks/rust.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        download_url = link.attrs['href']
+        checksum_url = download_url + '.sha256'
+        logger.debug("Found download URL: " + download_url)
+        logger.debug("Downloading checksum first, from " + checksum_url)
+
+        def checksum_downloaded(results):
+            checksum_result = next(iter(results.values()))  # Just get the first.
+            if checksum_result.error:
+                logger.error(checksum_result.error)
+                UI.return_main_screen()
```
Comment (didrocks) [inline umake/frameworks/rust.py]: Same, pass 1

File: umake/frameworks/rust.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                                                       checksum=Checksum(ChecksumType.sha256, checksum),
+                                                       ignore_encoding=True))
+            self.start_download_and_install()
+
+        DownloadCenter([DownloadItem(checksum_url)], on_done=checksum_downloaded, download=False)
+
+    def post_install(self):
+        """Add rust necessary env variables"""
+        add_env_to_user(self.name, {"PATH": {"value": os.path.join(self.install_path, "rustc", "bin")},
+                                    "PATH": {"value": os.path.join(self.install_path, "cargo", "bin")},
```
Comment (didrocks) [inline umake/frameworks/rust.py]: Here, you are creating a dictionary with two times the same key "PATH". That's the rason why you only see the last one being added, as the other doesn't exist (the key is unique in a map/dictionary).

So, you should rather pass:
{"PATH": {value: "{}:{}".format(os.path.join(self.install_path, "rustc", "bin"), os.path.join(self.install_path, "cargo", "bin"))}, …}

File: umake/frameworks/rust.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+                                    "LD_LIBRARY_PATH": {"value": os.path.join(self.install_path, "rustc", "lib")}})
```
Comment (didrocks) [inline umake/frameworks/rust.py]: Please add in the large tests a way a check that the LD_LIBRARY_PATH has been added.

File: umake/frameworks/rust.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        UI.delayed_display(DisplayMessage(_("You need to restart your current shell session for your {} installation "
+                                            "to work properly").format(self.name)))
+
+    @property
+    def is_installed(self):
```
Comment (didrocks) [inline umake/frameworks/rust.py]: This method shouldn't be required anymore with latest ubuntu make version.

**Comment by didrocks** [comment]
Hey jravetch! Thanks for this pull requests :)

This mostly looks good, I've added some comments and tried to answer the issue you are seeing with env variable.

On the first commits: I would say, let's try to keep the initial attribution to @Tinche as it's where the credit (and blame of course!) is due. I don't know if it's just github hiding the initial commit and your branched from https://github.com/jravetch/ubuntu-make/commit/0954ce93190749cd3cac69401032c8d030300277 (I didn't branch locally to check), but I think if it's not the case that you should start from this work and then fix user message + add tests (medium and large in separate commits, as you have done).

Also, it seems you didn't run the medium tests, right? (they can only fail for multiple reasons: one being the wrong inheritance, and the second one being that you didn't provide any mock assets and certificates like the other frameworks). Please do run them, in addition to the pep8 and the large tests. If you need help to get them running, do not hesitate to ask (you can only run one category or even one tests. Running the whole testsuite can be long ;)).

You need as well, as the Rust framework is overriding generic functionality of the BaseInstaller, to add more tests to ensure that errors on bad inputs are triggered as expect. Those are of course only mockable in the medium tests (as it's where we have the fake webserver and assets). Basically, in that case I would see:
- a 404 for download page
- an invalid download page content
- a 404 for checksum link
- an invalid checksum content

For the invalid contents, you can look at what is done for instance in tests/medium/test_ide.py:ArduinoIDEInContainer

(yeah, it seems that those 4 tests should be added as well to other test_ide.py frameworks, everytime we derived from BaseInstaller to do something particular, but that's a story for another PR).

Do not hesitate if anything isn't clear enough! I'm happy to give a hand :)
Thanks again!

**Comment by jravetch** [comment]
@didrocks Thanks very much for the comments.  Yeah I didn't branch properly from the original commit from @Tinche.  I'm gonna close this PR to start with a clean branch.

**Pull request closed without merging**